Zero input fraud proof

matteocoppola opened the pull request:
Simple fraud proof about the zero input rule.

Also changed the specs to reflect how we prove it.

matteocoppola commented:
I had to fix some errors that for some reason "aiken check" didn't find.
Also added the new fraud proof "min-fee"

GeorgeFlerovsky reviewed (changes requested):
The zero-input validator looks good, but min-fee is too simplified ;)

keyan-m commented on onchain/aiken/validators/fraud-proof/min-fee/step.ak:
```diff
@@ -0,0 +1,111 @@
+
+//TODO: This will need execution traces to calculate it
+fn get_min_transaction_fee(_transaction: MidgardTxCompact) {
+  0
```
We should use `todo` so that the compiler can catch it whenever we re-include failure on warnings in our CI.
```suggestion
  todo
```

keyan-m replied:
Oh we have `-D` included in our CI. @GeorgeFlerovsky what do you think if we remove it for the time being?

keyan-m commented:
Among all the implemented fraud proofs, "Missing Native Script" from this PR seems to be the only one that would require changes after we implement #298.

matteocoppola commented:
After last standup, the current direction is to limit the max number of unbounded elements like the inputs and outputs.
So Missing Native Script fraud proof can stay like I've done for now.

matteocoppola commented:
Added "missing signature" fraud proof

VenInf committed:
WIP: better error types

VenInf committed:
chore: format

VenInf committed:
WIP: attempt to insert deposits in the DBs

VenInf committed:
fix: date insertion

VenInf committed:
fix: date insertion

VenInf committed:
fix: comments, spacing

VenInf committed:
fix: db issues

VenInf committed:
WIP: wrong l2Address

VenInf committed:
fix: l2Address, rm logs

VenInf committed:
chore: format

keyan-m committed:
feat: add datatype for midgard addresses

keyan-m committed:
feat: incorporate `MidgardAddress` in `midgard-node`

keyan-m committed:
fix: add proper error message

VenInf committed:
fix: mv TransactionUnspentOutput converter and others

VenInf committed:
feat: applyDepositUTxOsToDatabases

VenInf committed:
feat: mv root calculations

VenInf committed:
fix: comments and logs

VenInf committed:
fix: comment

VenInf committed:
feat: deposit -> transaction cbor

keyan-m committed:
fix: conversions for midgard address

VenInf committed:
feat: rm unecessary checks in depositEventToCmlTransactionUnspentOutput

VenInf committed:
feat: + AddressHistoryDB insertions

VenInf committed:
chore: format

HarunJr committed:
add: extended always-succeeds for initializaton, abstracted linked-list and initialization example

VenInf committed:
fix: smaller tx.cbor

VenInf committed:
fix: rm not used line

keyan-m committed:
docs: minor formatting and comments

keyan-m committed:
refactor: minor rename

keyan-m committed:
refactor: more minor ones

VenInf committed:
fix: review issues

VenInf committed:
fix: extra log

VenInf committed:
WIP: reprodusing wierd tx.compose() behaviour

VenInf committed:
feat: fix package.json, revert debug logs

VenInf committed:
feat: proper insertion times

VenInf committed:
fix: insertionTime cleanup

VenInf committed:
chore: format

keyan-m committed:
feat: add retry schedules for lucid and config

HarunJr committed:
fix: updates as per review for the sdk

HarunJr committed:
update: node with improved sdk and fix reviews

HarunJr committed:
remove: init.ts

keyan-m committed:
feat: bring dev compose from `update-node-docs`

keyan-m committed:
refactor: better start log

AXELTOPOLINO committed:
Merge branch 'main' into matteoProofs

AXELTOPOLINO committed:
wip change from hashes to merkle roots

HarunJr committed:
update: review changes

HarunJr committed:
chore: remove redundant commented function

keyan-m committed:
Merge branch 'main' into basic-deposit

keyan-m committed:
Merge branch 'basic-deposit' into ledger-deposit

keyan-m committed:
Merge branch 'basic-deposit' into initialization-impl-only

HarunJr committed:
Resolve merge conflict: accept docker-compose.dev.yaml

HarunJr committed:
update midgard SDK and node according to review

HarunJr committed:
merge with remote lock file

HarunJr committed:
standardize lock file

HarunJr committed:
fix: rename mint and spend makeValidator results

keyan-m committed:
Merge branch 'main' into basic-deposit

keyan-m committed:
fix: le version and program execution

keyan-m committed:
fix: le version

keyan-m committed:
Merge branch 'basic-deposit' into ledger-deposit

keyan-m committed:
Merge branch 'basic-deposit' into initialization-impl-only

HarunJr committed:
refactor: add contract-specific init wrappers for linked lists

Replace generic helper calls with contract-specific wrappers that
encapsulate initialization logic and improve type safety.

HarunJr committed:
refactor(sdk): fix fraud proof catalogue initialization

- FraudProofCatalogue: singleton MPT root (not linked list)
- Add EMPTY_MPT_ROOT_HASH constant (Ethereum empty trie root)
- Update initialization to use correct MPT root hash

AXELTOPOLINO committed:
wip, need to update fraud proofs with merkle roots

HarunJr committed:
refactor: Parameterize MPT root and fix init logging

- SDK: Update initialization params to accept `mptRootHash`
- Node: Explicitly pass GENESIS_HASH_32 for initialization
- Node: Fix initialization result logging to show txHash

HarunJr committed:
fix: remove stateQueue idempotency until discussion

HarunJr committed:
refactor: simplify `initializeMidgard` return type

- Update `initializeMidgard` to return `txHash` string directly instead of an object wrapper.
- Update `getInitHandler` to consume the direct `txHash` return value.

HarunJr committed:
Refactor MPT root logic and sort validators

- Abstracted genesis UTxO extraction into a dedicated helper.
- Implemented in-memory MPT root calculation helper.
- Ensure validators are processed in lexicographical order by their script Hash.
- Optimized genesis UTxO generation to use concurrent hashing.
- Fixed error handling in listen command to properly catch specific errors.
- Unified UTxO batch operation logic.

HarunJr committed:
update: active operator and scripts in common.ts

HarunJr committed:
Refactor SDK init params and validator types

- Use `Pick` for HubOracleValidators to ensure type consistency.
- Update HubOracleInitParams to nest validators in a dedicated object.
- Remove unnecessary datum parameter from Scheduler initialization.
- Standardize rootData initialization to "00" for Registered and Retired operators.
- Update initialization transaction programs to align with new type definitions.

HarunJr committed:
feat: implement Validator Catalogue MPT root calculation

- refactor(genesis): replace `getGenesisUtxosFromValidators` with `getGenesisScriptInputs` to generate index-to-hash mappings for the validator catalogue
- feat(initialization): update `computeGenesisMptRoot` to use `keyValueMptRoot` with the new catalogue inputs
- chore(mpt): remove unused `computeRootFromUtxos` helper
- fix(listen): update error handling to match new MPT root calculation return types

HarunJr committed:
format: code

HarunJr committed:
fix: scheduler token name

AXELTOPOLINO committed:
pushing to merge with other branch and mantain consistency

AXELTOPOLINO committed:
merged main

HarunJr committed:
feat: implement fraud proof catalogue MPT root

Maps 4-byte indices to Blake2b-224 validator script hashes per spec §4.1.

Uses concurrent hashing with deterministic ordering defined in VALIDATOR_ORDER.
Tests verify pre-image retrieval, determinism, and 28-byte hash format.

Note: Validator ordering should be verified against on-chain code before production.

HarunJr committed:
refactor: address PR review feedback

- Complete hub oracle schema with all validator addresses and policy IDs
- Rename test file: mpt_roots.test.ts → fraud-proof-catalogue.test.ts
- Remove scheduler initialization datum
- Improve test validation to verify hash correctness
- Clean up test utilities (remove dotenv.config)

HarunJr committed:
fix: restore dotenv in tests and use path aliases

keyan-m committed:
feat: final touches to initialization

keyan-m committed:
Merge branch 'main' into basic-deposit

keyan-m committed:
Merge branch 'basic-deposit' into initialization-impl-only

keyan-m committed:
feat: checkpoint

keyan-m committed:
feat: finalize finishing touches to initialization

keyan-m committed:
refactor: repurpose sdk test utils

keyan-m committed:
fix: handle unspecified network for hub oracle

keyan-m committed:
Merge pull request #380 from Anastasia-Labs/initialization-impl-only

Expand the Initialization Process to Include All the Contracts

keyan-m committed:
fix: unspecified network error

keyan-m committed:
Merge branch 'basic-deposit' into ledger-deposit

matteocoppola committed:
currently broken, fixing conflicts

keyan-m committed:
Merge branch 'main' into basic-deposit

keyan-m committed:
Merge branch 'basic-deposit' into ledger-deposit

keyan-m committed:
refactor: use `AuthenticatedValidator` for hub

keyan-m committed:
refactor: 1st draft for reset using reduce

keyan-m committed:
refactor: reset refined by ai

keyan-m committed:
fix: dependency issues

keyan-m committed:
feat: handle block commitment edge case

Failure of submitting a block with 0 tx requests with deposit events did
not revert the changes applied to the ledger trie.

This commit implements the reversal logic as a potentionally temporary
solution.

keyan-m committed:
fix: propagate deposit reversion for other fails

matteocoppola committed:
refactoring of all fraud proofs after merging

keyan-m committed:
refactor: extract commitment utils

keyan-m committed:
Merge pull request #377 from Anastasia-Labs/ledger-deposit

Apply Deposit Events to Ledger

matteocoppola committed:
added no-reference-input fraud proofs

keyan-m committed:
fix: bug and inclusion time inclusivity

keyan-m committed:
fix: enforce libsodium version in `midgard-node`

Also fixed inclusion time inclusivity.

Drop-Table-Users committed:
WIP

martinceresa committed:
Adding Lean4 mechanization as a submodule

TotallyNotChase committed:
Update Ply for better errors

TotallyNotChase committed:
Add Haskell analogs for linked list datum type

TotallyNotChase committed:
Update script types

TotallyNotChase committed:
Fix improper root asset name for ActiveOperators init

TotallyNotChase committed:
Fix alwaysSucceedsUPLC

TotallyNotChase committed:
Update init contracts

TotallyNotChase committed:
Add initProtocol test

TotallyNotChase committed:
Hoist out initProtocol as a utility helper

Drop-Table-Users committed:
Going through files 1-by-1 + comments on where are similar fuel functions located; WIP transaction.ts left

TotallyNotChase committed:
Use reference scripts for init operators

TotallyNotChase committed:
Add initProtocol test scaffold

Drop-Table-Users committed:
Small refactoring

Drop-Table-Users committed:
Fix Mint type, add initial version of the Cardano <-> Midgard convertion functions

TotallyNotChase committed:
Add utilities and update registerOperator contract

TotallyNotChase committed:
Fix datum encoding

TotallyNotChase committed:
Declare full hub datum type

TotallyNotChase committed:
Fix hub oracle expectation

TotallyNotChase committed:
Add more utilities

TotallyNotChase committed:
Update constants

TotallyNotChase committed:
Fix registerOperators - needs Aiken to relax bond for minAda

TotallyNotChase committed:
Add register operator happy path test

Drop-Table-Users committed:
Use bigint to avoid losing precision

TotallyNotChase committed:
Minor cleanup

TotallyNotChase committed:
Rename RegisterOperator test file

TotallyNotChase committed:
Implement `findUtxoWithLink`

TotallyNotChase committed:
Relax required_bond checks to account for min ada requirements

TotallyNotChase committed:
WIP active operators

TotallyNotChase committed:
Some renaming; Utils; and fix script failure due to lack of addReference

keyan-m committed:
feat(aiken): start replacing keys of reg. ops.

keyan-m committed:
feat(aiken): update schedule with new reg. ops.

keyan-m committed:
feat(aiken): switching keys to act time seems done

keyan-m committed:
docs: activation times as keys of reg. ops.

keyan-m committed:
Merge pull request #427 from Anastasia-Labs/chase/aiken-relax-lovelace

Relax required_bond checks to account for min ada requirements

keyan-m committed:
Merge branch 'chase/aiken-relax-lovelace' into ordered-registered-ops

TotallyNotChase committed:
Add `Show` instances for `LinkedList` types

TotallyNotChase committed:
Fully working activateOperator and test

TotallyNotChase committed:
Add retireOperator contract

TotallyNotChase committed:
Add retireOperator test and consolidate tests

keyan-m committed:
feat(aiken): updt sched. dtm to covr empty act ops

Also brough over changes to `state-queue.ak` to relax validity range
requirements on block commitments.

See commit c4ff71d1725ad237a646e3319b16a3ee6facdff2

keyan-m committed:
feat(aiken): checkpoint

keyan-m committed:
feat(aiken): checkpoint

keyan-m committed:
feat(aiken): 1st draft for scheduler's skip logic

colll78 committed:
Merge pull request #423 from martinceresa/staging

Adding Lean4 mechanization as a submodule

Drop-Table-Users committed:
Forgot to push cardano-rountrip.test.ts

keyan-m committed:
docs: start drafting updated scheduler

keyan-m committed:
docs: progress with scheduler

keyan-m committed:
Merge branch 'update-linked-list-docs' into fix-scheduler

keyan-m committed:
feat: checkout

keyan-m committed:
Merge branch 'update-linked-list-docs' into ordered-registered-ops

keyan-m committed:
docs(opdir): specs seem complete

keyan-m committed:
docs: start drafting penalty box

keyan-m committed:
refactor(aiken): remove penalty box

Instead of dedicating another operator set, it seems better to extend
the retired ops. set to also house partially slashed operators.

keyan-m committed:
fix(aiken): minor errors

keyan-m committed:
feat(aiken): ensure scheduler is updated on slashs

keyan-m committed:
refactor(aiken): abstract op stat vs scheduler

keyan-m committed:
feat(aiken): extnd ret. endpt. 4 partial slashing

Giving strikes now does not care about the total count. The transfer
from active ops. to retired ops. now checks the inactivity strikes.

If few strikes are received, retirement requires operator's consent.
However, if the operator has received too many strikes, the transfer can
be performed by anyone, and it requires the slashing penalty to be paid
as transction fees. The rest must be preserved in the new retired node
(TODO, we probably want a separate parameter for this partial slashing).

The reason I've opted for not giving any rewards is that it seemed to
incentivized DDoS attacks on active ops. This way, only the Cardano
treasury benefits from the partial slashes.

keyan-m committed:
feat(aiken): overall refinements and fixes

keyan-m committed:
refactor(aiken): abstract slashing logic

keyan-m committed:
fix(aiken): resolve some compiler errors

keyan-m committed:
fix(aiken): resolve compile errors

keyan-m committed:
feat(aiken): progress drafting new scheduler

Drop-Table-Users committed:
Phase A validation wip

keyan-m committed:
feat(aiken): sched. cross val. in active ops.

keyan-m committed:
feat(aiken): time/consent validation 4 EndOfShift

keyan-m committed:
feat(aiken): disallow indefinite inactivity strkes

keyan-m committed:
feat(aiken): allow RewindSkip for single actv. op.

Drop-Table-Users committed:
phase-b based on tx-validation branch

keyan-m committed:
feat(aiken): extend comments a bit

keyan-m committed:
feat(aiken): add todos

keyan-m committed:
feat(aiken): `GoToNextDueToSkippedOperator`

keyan-m committed:
feat(aiken): 1 of sched's operator removal paths

colll78 committed:
Merge pull request #426 from Anastasia-Labs/feat/midgard-codec-lib

Midgard Custom Binary Encoding

keyan-m committed:
Merge branch 'main' into staging

keyan-m committed:
Merge branch 'staging' into ordered-registered-ops

keyan-m committed:
Merge branch 'staging' into fix-scheduler

keyan-m committed:
feat(aiken): check adv. approach in scheduler sync

keyan-m committed:
fix(aiken): add interval checks for skip endpoints

keyan-m committed:
fix(aiken): allow inactvt grace period for new ops

keyan-m committed:
fix(aiken): unused variable

keyan-m committed:
feat(aiken): add comments for scheduler helpers

keyan-m committed:
feat(aiken): `GoToNextDueToOperatorRemoval`

keyan-m committed:
feat(aiken): `AppointFirstOperator`

Also fixed rewind endpoints to check the registered operators set to
prove none are eligible for activation.

colll78 committed:
Merge pull request #425 from Anastasia-Labs/chase/offchain-operator-txs

Operator transactions + tests

keyan-m committed:
feat(aiken): 1st draft for re-reg. of a retired op

keyan-m committed:
Merge branch 'staging' into fix-scheduler

keyan-m committed:
fix(aiken): add missing argument for sched. helper

matteocoppola committed:
refactoring

keyan-m committed:
feat(aiken): datum getters for user events

matteocoppola committed:
renamed todo fraud proof no-output

matteocoppola committed:
merging staging and resolving conflicts

keyan-m committed:
feat(aiken): 1st draft for ignored user events

Simply looking at latest committed block's end time for inactivity
strikes was too restrictive. It meant operators were obliged to submit
block commitments even if they were empty.

By introducing a second timeout parameter, we can get more flexibility
in achieving a balance between imposing this kind of cost onto
operators, and guaranteeing liveness.

Here, we are allowing inactivity strikes by either pointing to a user
event's L1 UTxO to show that it's been slated for inclusion for long
enough (`operator_inactivity_timeout`) since last block commitment,
or skipping the operator if no blocks had been submitted for too long
(`max_inactivity_between_block_commitments`).

keyan-m committed:
docs: progress with the new scheduler

colll78 committed:
Merge pull request #431 from Anastasia-Labs/ordered-registered-ops

Make Registered Operators Set Ordered by Activation Times

keyan-m committed:
refactor(aiken): 1st drft abstracting ledger types

keyan-m committed:
fix(aiken): remove unneeded type

TotallyNotChase committed:
Fix output_header_hash calculation in state queue policy

keyan-m committed:
Merge pull request #437 from Anastasia-Labs/chase/state-queue-policy-fix

Fix output_header_hash calculation in state queue policy

Cauterization committed:
Preliminary off chain for tx orders (#361)

* feature: add tx order db

* feature: add latest tx order fetch time global

* feature: add tx order fetching fiber

* feature: add retrieveTimeBoundEntries to tx order db

* fix: add WAIT_BETWEEN_TX_ORDER_UTXO_FETCHES to config

* feature: add tx order event to commit-block-header

* fix: merge

* fix: getRoot method

* refactor: formatiting

* fix: remove processed mempool

* Revert "fix: remove processed mempool"

This reverts commit d620a1f6c6e3bcbf4b1dfbb6c2f1e2628eacfae8.

* fix: UserEventsUtils

* refactor: remove redundant comments and imports

* fix: revert commit block headers flow

* refactor: rename events to user events

* fix: revert commit block headers flow

* refactor: formatting

* add: tx order tx-builder

* update: removed commented code

* refactor: formatting

* fix: assetname calculation

* fix: removed `tx-builder.ts` file

* chore: formatting and removal of some files

* feat: export tx-orders from user-events

* add:CardanoTx for midgardTx

* Squashed commit of the following:

commit 2761db8b5160ff8ee74ba16f84f11f8e33f6fd38
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sun Nov 2 20:46:51 2025 +0100

    fix: better handling of edge case

    Available deposit events with no tx requests.

commit 29e269fb815230f25619a9b61bcec964b46fe34b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sun Nov 2 15:26:37 2025 +0100

    refactor: remove free vars from commitment worker

commit 1089490054ef899d4d668005339bd1292b31b088
Merge: 8374bc3 7ec65b8
Author: keyanm <86695999+keyan-m@users.noreply.github.com>
Date:   Sat Nov 1 22:09:01 2025 +0330

    Merge pull request #372 from Anastasia-Labs/reorganize-midgard-sdk

    Reorganize Midgard SDK

commit 7ec65b8e0e1a1a4e45b48239866dac816a25e00d
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sat Nov 1 19:36:46 2025 +0100

    fix: minor ones

commit e4be8496645a9b9d291e5e9ebeb745e9078dbf94
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:36:14 2025 +0300

    refactor: formatting

commit c5ee4f03782d3ec410d5533d9cda00affcf4bf47
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:34:37 2025 +0300

    fix: midgard-node imports

commit 8fea8c6276c1475cd063b95a30ff43d2704b7b7b
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:31:12 2025 +0300

    fix: error type in unsignedInitStateQueueTxProgram

commit f2f8bb744b61843f4dfaa6338088f6b80b6ec240
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:26:20 2025 +0300

    fix: error type in unsignedInitStateQueueTxProgram

commit 6ea32a905274fa6454ec89bbd945ed857cb1b9fb
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:00:00 2025 +0300

    refactor: merge utils.ts and common.ts

commit 7300f9a0e204a32f325db68dbc7de5a5833013b6
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:47:07 2025 +0300

    refactor: remove consensus folder

commit e2809a67611a30d888dcae51b46a4ca10b628601
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:44:38 2025 +0300

    refactor: merge core.ts with common.ts

commit 6c6c8482e32d77d04e02e6f5225612bf935ef565
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:15:33 2025 +0300

    refactor: utils

commit 5f1c62954276dbc576898b569634b231129c5bdb
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:41:38 2025 +0300

    refactor: index modules

commit b9f02774355091d2de57824081d0c32ce17b14a2
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:36:09 2025 +0300

    refactor: fraud proof computation thread

commit dc07d2e4ae4258c7662e5132f7b2276c07228d1b
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:31:39 2025 +0300

    refactor: fraud proof tokens

commit 544bde7b5b4d478551ded7503afcd74c4806c993
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:28:08 2025 +0300

    refactor: fraud proof catalogue

commit a65a756e15cbc4381baa4ed8251bd8d7322090f5
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:21:49 2025 +0300

    fix: scheduler init tx name

commit 1c40d2d03c0f02e05f7d9c078d7a0fed1aa35043
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:55:51 2025 +0300

    refactor: retired operators

commit f98cab18b071707ee5a70c40c01e2ec9158d3cc3
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:42:44 2025 +0300

    refactor: registered operators

commit 222e38d6c6c4d7d0f4c8400d2a5cd5fe460d24c9
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:26:37 2025 +0300

    refactor: active operator

commit 113f87ad016c5b025e4993d482cbf4439f19308e
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:10:32 2025 +0300

    refactor: scheduler

commit b026c86dfa828b30baf7a011a016477f35acd046
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:08:45 2025 +0300

    refactor: formatting

commit e57a172a002d60e77a1f7dd02f652a0ae42701cf
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:05:13 2025 +0300

    refactor: clean imports

commit c4801bccd0269198f265cce4f2c4d048f56f2641
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:04:31 2025 +0300

    refactor: merge ledger-state with state-queue

commit 8fdb974b232077078acb8b1be8eadb9855c38944
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:59:39 2025 +0300

    refactor: hub oracle

commit c833b30f44160372d0884881da47c13d2e2096ca
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:49:47 2025 +0300

    refactor: clean user events and state queue exports

commit 31e23d01f2312b87a43a71a4b438bd9437807510
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:46:59 2025 +0300

    refactor: state queue

commit 16c165d4d69ebe31e5a6282b23f07aba3a8d64a3
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:51:01 2025 +0100

    feat: update sdk's index

commit 08b9487b7fcc00176f51265539cd16d13e1c296d
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:18:45 2025 +0100

    chore: minor fix and formatting

commit 8bf5d2685941ee718eae1e89c6881b698edc4ada
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:14:53 2025 +0100

    refactor: initial step

commit 8374bc3d5503fe57a30711f9ccdd976d3bfcce98
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 23:24:13 2025 +0100

    feat: comments and logs

commit 7eefbc937e7229400be73dff8a977a4ce9dda72a
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 22:08:16 2025 +0100

    refactor: better empty root helper

commit 20c138c07a0c94a3db61a4a5406b246d6457ff4b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 21:55:42 2025 +0100

    feat: cover edge case of 0 mempool txs with events

commit 3c26dbe9c0a82832f196e58243812803d0436a7b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:34:35 2025 +0100

    feat: add a retry policy for genesis deposits

commit 4c24bcf8574299ae1391489287fb993e274ec9a7
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:16:39 2025 +0100

    feat: updated initial start time for deposits

commit ad1da27e9e579e5a80a29c53f9edcc12982d19cc
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:00:38 2025 +0100

    feat: update `midgard-node` genesis

commit c7a79ca7dcd1317e1396271ba4b66ba652b5a253
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 09:54:10 2025 +0100

    feat: add deposit amount to deposit params

    Also removed `inclusionTime` and let the tx builder pick it by adding
    the wait duration protocol parameter to the current time.

* fix: merge

* fix: sdk namespaces

* Squashed commit of the following:

commit be446b2672b803d6dbace40ef86e6e0238df804a
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Tue Nov 4 15:58:51 2025 +0100

    fix: wallet selection, order of start/end time etc

commit f64fafeb23d59ac3af4531b82a3806f5df274d28
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Tue Nov 4 13:22:21 2025 +0100

    refactor: rename module

commit 2761db8b5160ff8ee74ba16f84f11f8e33f6fd38
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sun Nov 2 20:46:51 2025 +0100

    fix: better handling of edge case

    Available deposit events with no tx requests.

commit 29e269fb815230f25619a9b61bcec964b46fe34b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sun Nov 2 15:26:37 2025 +0100

    refactor: remove free vars from commitment worker

commit 1089490054ef899d4d668005339bd1292b31b088
Merge: 8374bc3 7ec65b8
Author: keyanm <86695999+keyan-m@users.noreply.github.com>
Date:   Sat Nov 1 22:09:01 2025 +0330

    Merge pull request #372 from Anastasia-Labs/reorganize-midgard-sdk

    Reorganize Midgard SDK

commit 7ec65b8e0e1a1a4e45b48239866dac816a25e00d
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Sat Nov 1 19:36:46 2025 +0100

    fix: minor ones

commit e4be8496645a9b9d291e5e9ebeb745e9078dbf94
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:36:14 2025 +0300

    refactor: formatting

commit c5ee4f03782d3ec410d5533d9cda00affcf4bf47
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:34:37 2025 +0300

    fix: midgard-node imports

commit 8fea8c6276c1475cd063b95a30ff43d2704b7b7b
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:31:12 2025 +0300

    fix: error type in unsignedInitStateQueueTxProgram

commit f2f8bb744b61843f4dfaa6338088f6b80b6ec240
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:26:20 2025 +0300

    fix: error type in unsignedInitStateQueueTxProgram

commit 6ea32a905274fa6454ec89bbd945ed857cb1b9fb
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 17:00:00 2025 +0300

    refactor: merge utils.ts and common.ts

commit 7300f9a0e204a32f325db68dbc7de5a5833013b6
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:47:07 2025 +0300

    refactor: remove consensus folder

commit e2809a67611a30d888dcae51b46a4ca10b628601
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:44:38 2025 +0300

    refactor: merge core.ts with common.ts

commit 6c6c8482e32d77d04e02e6f5225612bf935ef565
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 16:15:33 2025 +0300

    refactor: utils

commit 5f1c62954276dbc576898b569634b231129c5bdb
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:41:38 2025 +0300

    refactor: index modules

commit b9f02774355091d2de57824081d0c32ce17b14a2
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:36:09 2025 +0300

    refactor: fraud proof computation thread

commit dc07d2e4ae4258c7662e5132f7b2276c07228d1b
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:31:39 2025 +0300

    refactor: fraud proof tokens

commit 544bde7b5b4d478551ded7503afcd74c4806c993
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:28:08 2025 +0300

    refactor: fraud proof catalogue

commit a65a756e15cbc4381baa4ed8251bd8d7322090f5
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 15:21:49 2025 +0300

    fix: scheduler init tx name

commit 1c40d2d03c0f02e05f7d9c078d7a0fed1aa35043
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:55:51 2025 +0300

    refactor: retired operators

commit f98cab18b071707ee5a70c40c01e2ec9158d3cc3
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:42:44 2025 +0300

    refactor: registered operators

commit 222e38d6c6c4d7d0f4c8400d2a5cd5fe460d24c9
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:26:37 2025 +0300

    refactor: active operator

commit 113f87ad016c5b025e4993d482cbf4439f19308e
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:10:32 2025 +0300

    refactor: scheduler

commit b026c86dfa828b30baf7a011a016477f35acd046
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:08:45 2025 +0300

    refactor: formatting

commit e57a172a002d60e77a1f7dd02f652a0ae42701cf
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:05:13 2025 +0300

    refactor: clean imports

commit c4801bccd0269198f265cce4f2c4d048f56f2641
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 14:04:31 2025 +0300

    refactor: merge ledger-state with state-queue

commit 8fdb974b232077078acb8b1be8eadb9855c38944
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:59:39 2025 +0300

    refactor: hub oracle

commit c833b30f44160372d0884881da47c13d2e2096ca
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:49:47 2025 +0300

    refactor: clean user events and state queue exports

commit 31e23d01f2312b87a43a71a4b438bd9437807510
Author: Grigory Petrov <71918928+Cauterization@users.noreply.github.com>
Date:   Sat Nov 1 13:46:59 2025 +0300

    refactor: state queue

commit 16c165d4d69ebe31e5a6282b23f07aba3a8d64a3
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:51:01 2025 +0100

    feat: update sdk's index

commit 08b9487b7fcc00176f51265539cd16d13e1c296d
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:18:45 2025 +0100

    chore: minor fix and formatting

commit 8bf5d2685941ee718eae1e89c6881b698edc4ada
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Fri Oct 31 14:14:53 2025 +0100

    refactor: initial step

commit 8374bc3d5503fe57a30711f9ccdd976d3bfcce98
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 23:24:13 2025 +0100

    feat: comments and logs

commit 7eefbc937e7229400be73dff8a977a4ce9dda72a
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 22:08:16 2025 +0100

    refactor: better empty root helper

commit 20c138c07a0c94a3db61a4a5406b246d6457ff4b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 21:55:42 2025 +0100

    feat: cover edge case of 0 mempool txs with events

commit 3c26dbe9c0a82832f196e58243812803d0436a7b
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:34:35 2025 +0100

    feat: add a retry policy for genesis deposits

commit 4c24bcf8574299ae1391489287fb993e274ec9a7
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:16:39 2025 +0100

    feat: updated initial start time for deposits

commit ad1da27e9e579e5a80a29c53f9edcc12982d19cc
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 10:00:38 2025 +0100

    feat: update `midgard-node` genesis

commit c7a79ca7dcd1317e1396271ba4b66ba652b5a253
Author: keyanmaskoot <keyanmaskoot@gmail.com>
Date:   Thu Oct 30 09:54:10 2025 +0100

    feat: add deposit amount to deposit params

    Also removed `inclusionTime` and let the tx builder pick it by adding
    the wait duration protocol parameter to the current time.

* fix: merge

* fix: deposit event helper

* chore: merge with tx-builder-tx-order

* feature: add missing entries to sdk tx order

* add:withdrawal types

* add:tx order always succeeds contracts

* chore: build always succeeds

* feature: tx order addressHisroryDB, add block commision to no tx request case

* feature: genesis tx order tx

* fix: use hashes instead of full tx body and wits in incompleteTxOrderTxProgram

* fix: init LATEST_TX_ORDER_FETCH_TIME

* fix: change tx order datum format

* update: abstract mintredeemer

* merge tx-order

* feature: add helper to parse address data

* add:abstraction for mint transaction

* add:CardanoDatum

* add: abstract mint tx for deposit and tx order

* add:withdrawal validator to AlwaysSucceedsContract

* chore: build always succeeds aiken project

* fix:inconsistent inclusionTime on tx-order and withdrawal

* fix:changed validTo type to number

* feature: update withdrwal order in sdk to match other user events

* refactor: formatting

* refactor: formatting

* feat: add withdrawals fetch fiber

* Revert "feat: add withdrawals fetch fiber"

This reverts commit e1582ebfc1fbefc3553cc0da388d3f6979eb75c6.

* feat: add withdrawals tx root to commit block headers

* feature: add utxo set update to withdrawals,deposits and tx orders

* feature: add utxo set update to tx requests

* feature: add size of processed txs to user actions

* fix: tx order params in genesis

* feature: add addresshistory db update to user actions

* refactor: formatting

* add:schemas for settlement contract components

* add: extract settlement utxo without claim

* add: extract active operator utxo

* add:reference hub oracle utxo

* fix: removed redundant

* fix: removed redundant fn utxosAtByNFTPolicyId

* add:HubOracleUTxO and SchedulerUTxO

* fix: abstract function for reference utxos

* fix:complete attach resolution tx

* fix: move refUTxO selection logic into fetch function

* add:disprove initial draft

* add:fetch user Event UTxOs

* Revert "Merge remote-tracking branch 'origin/Preliminary-Off-Chain-for-Withdrawals' into tx-builder-settlement-contract"

This reverts commit 69238923a2003c5bd9c4ab787e25c92bcc467bc9, reversing
changes made to 2c245fb0bd2ad0c5a96636f6e2782c659c4c6da8.

* add: remove bad operator transaction

* add:moved activeoperator logic

* fix:corrected txUpperBound

* add:func getOperatorStatus

* fix:Used Effect.try in huboracleDatum extraction

* fix: added ValidityOverride for EventType

* fix:selection logic of usereventutxos

* fix:used Effect.forEach

* add:resolve endpoint

* fix:revised logic to check userevent utxos are resolved

* fix:used forEach for consistency

* fix: getOperatorNodeUtxo logic changed

* fix:modified logic for fetchUserEventUtxos

* formatted

* fix: added function utxosToSettlementUTxOs

* fix:created  and moved huboracle and scheduler related funcs

* add: abstract function for utxosToUtxos in common.ts

* fix: added abstract function for userevents

* fix:renamed abstract function and copied over withdrawals latest changes

* fix: tx programs uses utxosToAuthenticUTxOs

* fix:provided SettlementUTxO via function paramater

* fix: added SettlementUTxO into params

* fix:removed invalid operator status

* refactor(sdk): refine user events

* refactor(node): unify user event fibers

* feat(node): shorter names for user events fiber

* feat(node): withdrawals, table and incl. in fiber

* fix(node): prepare genesis tx order for completion

* feat(node): genesis depost and tx-order in 1 tx

* refactor(sdk): sdk authentic utxo internals

* fix(sdk): scheduler utxo

* refactor(sdk): generic fetch for single-utxo vals

* refactor(sdk): reorganize files

* feat(demo): bump pnpm

* fix(node): genesis user events tx

* feat(node): user events ledger application helpers

---------

Co-authored-by: Suganya Arun <realsuganya01@gmail.com>
Co-authored-by: keyanmaskoot <keyanmaskoot@gmail.com>
Co-authored-by: keyanm <86695999+keyan-m@users.noreply.github.com>

keyan-m committed:
Queue Block Submission in `midgard-node` (#424)

* feat(node): new table for unsubmitted blocks

* feat(node): 1st draft for reformed confrmtion wrkr

* feat(sdk): cheaper header hash extraction

* feat(node): reform confirmation worker

As a step towards caching unsubmitted block commitment transactions,
this commit removes the confirmation worker, and adds the block
submission fiber that submits the oldest unsubmitted block and keeps the
two global variable in sync.

* chore: small refinement

* feat(node): retrieve latest unsubmitted blk, drft1

* fix(node): use `NotFoundError` in BlocksDB

* refactor(node): more effectful style

* fix(node): dynamic columns in sql query

* feat(node): stat columns to UnsubmittedBlocksDB

* refactor(node): mempool trie -> txs trie

* feat(node): 1st draft of withdrawal application fn

* chore(node): move utxo list ser/deser functions

* refactor(node): simplify block submission fiber

* feat(node): add block interval to db

* chore(node): rename variable

* fix(aiken): permissive validity range block cmtmnt

In block commitment, requiring the upper bound of the submitted block to
be equal to its Cardano transaction is both unnecessary, and prohibitive
for `midgard-node`.

In `midgard-node` we need to first establish the block's upper bound
before being able to find its roots, which is a relatively time
consuming operation. It is also not possible to submit a transaction
which its upper bound is in the past.

So the node first locks in the block's upper bound, applies withdrawal
events, followed by tx orders, tx request and deposits in order.

Finally the signed transaction is put in a queue to be submitted later.
Since LE is not capable of submitting multiple transactions
simultaneously (at least not seemingly), the wait between confirmation
of each block commitment tx can also add to the time between the
of the transaction and submitting it.

* feat(docs): document process flow for blck cmtmnt

* docs(node): minor refinements

* feat(node): progress impl block submission logic

* fix: minor typo

* feat(node): 1st draft of block submission fiber

* chore: formatting

* feat(node): add submitted block to `BlocksTxsDB`

`UnsubmittedBlocksDB` is to be renamed to `BlocksDB`, and the current
`BlocksDB` is to be renamed to `BlocksTxsDB`.

* chore(node): compile errors and better names

* chore(node): rename fiber

* chore(node): rename `BlocksDB` to `BlocksTxsDB`

* chore(node): rename `UnsubmittedBlocksDB`

`UnsubmittedBlocksDB` -> `BlocksDB`

* chore(node): better database module names

* chore(node): reorganizing further

* feat(node): add dedicated op wallet 4 blck cmtmnt

* chore(node): better config variable names

* refactor(node): send some helpers to db modules

* feat(node): `BlocksDB` helper

* feat(node): progress with cmtmnt & upd addr hstry

* feat(node): prpr `AddressHistoryDB` insertion logc

* docs(node): start docs for database tables

* docs(node): finalize 1st draft for database tables

* docs(node): fix typo

* docs(node): better wording

* feat(node): addr history insertion for withdrawals

* feat(node): avoid addr hist insertion on discovery

Injecting corresponding `AddressHistoryDB` entry upon discovery of user
events leads to complications for withdrawal and/or tx order events that
spend non-existent inputs.

Querying user events tables on demand for discovering involved addresses
seems like a reasonable compromise (at least for now).

* fix(node): withdrawal addr history entry

* refactor(node): better readability

* feat(node): addr history tx order insertion

* feat(node): better coordination between modules

* feat(node): block submission fiber 1s draft

Block submission seems to be covering almost all edger cases at this
point. Failures might need handling still.

* fix(node): retrieval of earliest unsubmitted block

* docs(node): align with implemented submission code

* feat(node): progress with block commitment

Also removed the "committed" status from `AddressHistoryDB` since
updating this table at the time of block commitment (which is the most
performance sensitive part of the process) would add lots of database
overhead.

* feat(node): revisit mpt ledger application helpers

* feat(node): ensure ascending ordering of events

* feat(node): appended block retrieval from `BlocksDB`

* feat(node): new block commitment logic

* feat(node): new ver. of block commitment worker

Not tested yet.

* fix(node): resolve compile errors

* docs(node): minor refinements

* feat(node): minor cleanup

* fixes: crate table SQL; genesis state

* Update the log to also show the number of L2 TXs being processed

* Resolve upsertEntries TODO

---------

Co-authored-by: Drop-Table-Users <avalach84@gmail.com>

keyan-m committed:
docs: 1st draft for max allowed inactivity period

keyan-m committed:
feat(aiken): some refinements

keyan-m committed:
docs: 1st draft for 4 out of 6 sched. redeemers

keyan-m committed:
refactor(aiken): reuse helper in scheduler

keyan-m committed:
docs: 1st draft for remaining 3 sched. endpoitns

keyan-m committed:
Merge branch 'staging' into fix-scheduler

keyan-m committed:
fix(aiken): resolve compile errors/warnings

keyan-m committed:
fix(aiken): full slash of partially slashed ops.

Updates the slashing helper to cover the full slashing of partially
slashed operators at the registered operators set.

There are currently two protocol parameters for the slashing penalty:
- `slashing_penalty` is the portion of bond that goes to the Cardano
  treasury when slashing fraudulent operators.
- `inactivity_slashing_penalty` is a similar parameter, but dedicated to
  operators with enough inactivity strikes.

Since both are paid as tx fees, this results in the implicit requirement
of `slashing_penalty` being larger than `inactivity_slashing_penalty` in
order to preserve fraud prover rewards for both cases.

keyan-m committed:
feat(aiken): add slashing from reg. ops. set

Now that it's allowed for retired operators to re-register and transfer
their nodes into the registered operators set (which carries over their
`bond_unlock_time` values), it must also be possible to slash those who
are proved to be fraudulent from the registered set.

This commit adds this feature, while also refining the abstractions
around the slashing helper. Namely, removing the requirement for syncing
with the scheduler. Now a dedicated helper is defined for retired and
registered sets that won't require involvement of the scheduler UTxO.

keyan-m committed:
docs(aiken): refine comments

keyan-m committed:
docs: progress revising operator directory specs

keyan-m committed:
feat(aiken): add todo

keyan-m committed:
fix(aiken): slash fraudulent ops. from reg. list

keyan-m committed:
docs: move inactivity criteria to op. directory

keyan-m committed:
refactor(aiken): move actve helper 2 vals from lib

keyan-m committed:
refactor(aiken): better naming

keyan-m committed:
docs: revised active operators

keyan-m committed:
docs: specs seems to be fully up to date with code

keyan-m committed:
chore(aiken): bump dependency version

keyan-m committed:
feat(aiken): remove some todos

Revisited some parts which had ambiguities with regards to the
possibility of referencing spending inputs, which is not allowed in
Plutus V3.

keyan-m committed:
docs: align with code

keyan-m committed:
chore(aiken): refine comment

keyan-m committed:
fix(ci): attempt

keyan-m committed:
feat(aiken): checkpoint

keyan-m committed:
Merge branch 'fix-scheduler' into matteoProofs

keyan-m committed:
Merge branch 'matteoProofs' into general-fixes

keyan-m committed:
feat(aiken): checkpoint

keyan-m committed:
refine(aiken): double spend fraud proof

keyan-m committed:
refactor(aiken): start revising invalid sig. fp.

keyan-m committed:
feat(aiken): extend fp. helper for othe preimages

`pass_tx_to_next_step` proved too restrictive as it only worked for
fraud proofs that needed access to the transaction's body.

This commit extends the interface so that fraud proofs that only need
transaction's witness set, or both the body and witness set can also
utilize it.

keyan-m committed:
refactor(aiken): conform a few of fraud proofs

keyan-m committed:
refactor(aiken): invalid sig. fraud prf seems done

keyan-m committed:
refactor(aiken): minor changes to min fee fp.

keyan-m committed:
refactor(aiken): missing native script fraud proof

keyan-m committed:
refine(aiken): fraud proof parameters

keyan-m committed:
refine(aiken): missing required signer fraud proof

keyan-m committed:
refactor(aiken): no reference input fraud proof

* Added cause for absence of the reference input: either already missing
  from block's prev ledger, or spent in a withdrawal event. This
  required one additional step (TODO: not reflected in specs yet).
* Added an extra step based on the specs that disallows the reference
  input to be produced in the same block (TODO? This doesn't _seem_
  ideal. However, even if we distinguish tx orders from tx requests,
  it'd take complete traversal of its tree to show there are no txs that
  produce the missing ref input).

keyan-m committed:
docs: withdrawn reference input fraud proof

My refactor of `no-reference-input` implements two fraud proofs:
`NO-REFERENCE-INPUT` and `WITHDRAWN-REFERENCE-INPUT` (which is the fraud
proof added to specs in this commit).

This should be split in two in a subsequent commit/PR.

keyan-m committed:
fix(aiken): resolve compile errors

keyan-m committed:
fix(aiken): refine comments

keyan-m committed:
fix(aiken): remove `withdrawn-reference-input`

To be added back as a separate fraud proof in next commit.

keyan-m committed:
feat(aiken): withdrawn reference fraud proof

keyan-m committed:
fix(aiken): rename fraud proof to align with specs

keyan-m committed:
chore(aiken): remove artifact

keyan-m committed:
Merge branch 'matteoProofs' into general-fixes

keyan-m committed:
refactore(aiken): use type alias

keyan-m committed:
refine(aiken): reuse type aliases

colll78 committed:
Merge pull request #441 from Anastasia-Labs/general-fixes

Refine Ledger Types and Fraud Proofs

keyan-m commented:
Closing in favor of #434
